components/server/src/api/server.ts

```diff
@@ -217,40 +217,50 @@ export class API {
                         }
                     };
 
-                    const apply = async <T>(): Promise<T> => {
-                        const subjectId = await self.verify(context);
-                        const isAuthenticated = !!subjectId;
+                    // actually call the RPC handler
+                    const auth = async () => {
+                        // Authenticate
+                        const userId = await self.verify(connectContext);
+                        const isAuthenticated = !!userId;
                         const requiresAuthentication = !Unauthenticated.get(target, prop);
-
                         if (!isAuthenticated && requiresAuthentication) {
                             throw new ConnectError("unauthenticated", Code.Unauthenticated);
                         }
 
                         if (isAuthenticated) {
-                            await rateLimit(subjectId);
-                            context.user = await self.ensureFgaMigration(subjectId);
+                            await rateLimit(userId);
+                            await self.ensureFgaMigration(userId);
                         }
-
                         // TODO(at) if unauthenticated, we still need to apply enforece a rate limit
 
-                        // actually call the RPC handler
+                        return userId ? SubjectId.fromUserId(userId) : undefined;
+                    };
+
+                    const apply = async <T>(): Promise<T> => {
                         return Reflect.apply(target[prop as any], target, args);
                     };
                     if (grpc_type === "unary" || grpc_type === "client_stream") {
                         return withRequestContext(async () => {
                             try {
-                                const promise = await apply<Promise<any>>();
-                                const result = await promise;
+                                const subjectId = await auth();
+                                const result = await runWithSubjectId(subjectId, async () => {
+                                    const promise = await apply<Promise<any>>();
+                                    return await promise;
+                                });
                                 done();
                                 return result;
                             } catch (e) {
                                 handleError(e);
                             }
                         });
                     }
+
+                    let subjectId: SubjectId | undefined = undefined;
                     return wrapAsyncGenerator(
                         (async function* () {
                             try {
+                                subjectId = await auth();
+
                                 const generator = await apply<AsyncGenerator<any>>();
```

Comment:
I'm not 100% sure it will be exected with subjectId, but we will learn it in PRs there we make use of streamin APIs and can fix there.

Reply:
Although I assume you tested with dummy service and results there returned a proepr subhect id

Reply:
I will re-do that. Last time might have been with the previous implementation. :grimacing: :+1:

Reply:
There seems to be an issue indeed. Trying to debug it...

Reply:
So... it took me 2h straight to debug, fix and test in the preview. It crashed multiple times, somehow, and in the end I could not access it anymore, at all. :see_no_evil: 
Just triggered a re-deploy, should work then! :crossed_fingers: